File: app/src/main/java/com/vswitch/configure/WifiCredentialsClient.kt
package com.vswitch.configure

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Calls the ESP32 REST endpoint:
 * POST http://192.168.4.1:8080/wifi-credentials
 * Body: {"ssid":"<home wifi ssid>","password":"<home wifi password>"}
 */
class WifiCredentialsClient(
    private val port: Int = 8080,
    private val gatewayHost: String = "192.168.4.1"
) {

    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .build()

    sealed class Result {
        data object Success : Result()
        data class HttpError(val code: Int) : Result()
        data class NetworkError(val message: String) : Result()
    }

    fun configureWifi(
        homeWifiSsid: String,
        homeWifiPassword: String,
        deviceSerialNumber: String
    ): Result {
        val payload = JSONObject()
            .put("ssid", homeWifiSsid)
            .put("password", homeWifiPassword)
            .toString()

        val requestBody = payload.toRequestBody(JSON_MEDIA_TYPE)
        val hosts = listOf(gatewayHost, "$deviceSerialNumber.local")

        var lastHttpCode = -1
        var lastNetworkError: String? = null

        for (host in hosts) {
            val request = Request.Builder()
                .url("http://$host:$port/wifi-credentials")
                .post(requestBody)
                .header("Content-Type", "application/json")
                .build()

            try {
                httpClient.newCall(request).execute().use { response ->
                    if (response.isSuccessful) {
                        return Result.Success
                    }
                    lastHttpCode = response.code
                }
            } catch (error: IOException) {
                lastNetworkError = error.message ?: error.javaClass.simpleName
            }
        }

        return when {
            lastHttpCode > 0 -> Result.HttpError(lastHttpCode)
            else -> Result.NetworkError(lastNetworkError ?: "Unknown network error")
        }
    }

    companion object {
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }
}

File: water_meter_app/lib/core/provisioning/wifi_credentials_client.dart
import 'package:dio/dio.dart';

sealed class WifiConfigureResult {
  const WifiConfigureResult();
}

class WifiConfigureSuccess extends WifiConfigureResult {
  const WifiConfigureSuccess();
}

class WifiConfigureHttpError extends WifiConfigureResult {
  const WifiConfigureHttpError(this.code);
  final int code;
}

class WifiConfigureNetworkError extends WifiConfigureResult {
  const WifiConfigureNetworkError(this.message);
  final String message;
}

class WifiCredentialsClient {
  WifiCredentialsClient({Dio? dio, this.port = 8080, this.gatewayHost = '192.168.4.1'})
      : _dio = dio ??
            Dio(
              BaseOptions(
                connectTimeout: const Duration(seconds: 15),
                receiveTimeout: const Duration(seconds: 15),
                sendTimeout: const Duration(seconds: 15),
              ),
            );

  final Dio _dio;
  final int port;
  final String gatewayHost;

  Future<WifiConfigureResult> configureWifi({
    required String homeWifiSsid,
    required String homeWifiPassword,
    required String deviceSerialNumber,
  }) async {
    final payload = {'ssid': homeWifiSsid, 'password': homeWifiPassword};
    final hosts = [gatewayHost, '$deviceSerialNumber.local'];

    var lastHttpCode = -1;
    String? lastNetworkError;

    for (final host in hosts) {
      try {
        final response = await _dio.post<Map<String, dynamic>>(
          'http://$host:$port/wifi-credentials',
          data: payload,
          options: Options(contentType: 'application/json'),
        );
        if (response.statusCode != null &&
            response.statusCode! >= 200 &&
            response.statusCode! < 300) {
          return const WifiConfigureSuccess();
        }
        lastHttpCode = response.statusCode ?? -1;
      } on DioException catch (error) {
        if (error.response?.statusCode != null) {
          lastHttpCode = error.response!.statusCode!;
        } else {
          lastNetworkError = error.message ?? error.type.name;
        }
      } catch (error) {
        lastNetworkError = error.toString();
      }
    }

    if (lastHttpCode > 0) {
      return WifiConfigureHttpError(lastHttpCode);
    }
    return WifiConfigureNetworkError(lastNetworkError ?? 'Unknown network error');
  }
}
